rust/otap-dataflow/crates/otap/src/experimental/geneva_exporter/mod.rs

```diff
@@ -0,0 +1,460 @@
+            auth_method,
+            tenant: config.tenant.clone(),
+            role_name: config.role_name.clone(),
+            role_instance: config.role_instance.clone(),
+            msi_resource,
+        };
+
+        // Initialize Geneva client
+        let geneva_client = GenevaClient::new(client_config).map_err(|e| {
+            otap_df_config::error::Error::InvalidUserConfig {
+                error: format!("Failed to initialize Geneva client: {}", e),
+            }
+        })?;
+
+        Ok(Self {
+            config,
+            metrics,
+            geneva_client,
+        })
+    }
+
+    /// Get exporter configuration
+    #[must_use]
+    pub fn config(&self) -> &Config {
+        &self.config
+    }
+
+    /// Handle PData message by converting OTAP to OTLP and uploading to Geneva
+    ///
+    /// TODO: Ideal Approach - Geneva uploader should accept OTAP directly
+    ///
+    /// Current implementation converts OTAP → OTLP → Geneva format, which has overhead:
+    /// - OTAP (Arrow RecordBatch) → OTLP (protobuf) conversion step
+    /// - OTLP (protobuf) → Geneva format conversion
+    ///
+    /// Ideal implementation would be:
+    /// - OTAP (Arrow RecordBatch) → Geneva format directly
+    ///
+    /// This requires modifying the geneva-uploader crate to accept Arrow RecordBatch
+    /// and process it directly, rather than going through OTLP ResourceLogs/ResourceSpans.
+    async fn handle_pdata(
+        &self,
+        pdata: OtapPdata,
+        effect_handler: &EffectHandler<OtapPdata>,
+    ) -> Result<(), String> {
+        // Split pdata into context and payload
+        let (_context, payload) = pdata.into_parts();
+
+        // Convert OTAP payload to OTLP bytes
+        // TODO: This conversion step should be eliminated (see method documentation above)
+        let otlp_bytes: OtlpProtoBytes = payload
+            .try_into()
+            .map_err(|e| format!("Failed to convert OTAP to OTLP: {:?}", e))?;
+
+        // Process based on signal type
+        match otlp_bytes {
+            OtlpProtoBytes::ExportLogsRequest(bytes) => {
+                effect_handler
+                    .info("Converting and uploading logs to Geneva")
+                    .await;
+
+                // Decode OTLP bytes to ResourceLogs
+                let logs_request = ExportLogsServiceRequest::decode(&bytes[..])
+                    .map_err(|e| format!("Failed to decode logs request: {}", e))?;
+
+                // Encode and compress using Geneva client
+                let batches = self
+                    .geneva_client
+                    .encode_and_compress_logs(&logs_request.resource_logs)
+                    .map_err(|e| format!("Failed to encode logs: {}", e))?;
+
+                // Upload each batch
+                for batch in batches {
+                    self.geneva_client
+                        .upload_batch(&batch)
+                        .await
```

Comment:
This will be sequential export. Add a TODO to check if we want to change this later.